components/eam/src/physics/cam/cam_diagnostics.F90

```diff
@@ -2393,4 +2412,320 @@ subroutine diag_state_b4_phys_write (state)
+
+      real(r8), intent(in) :: q(pcols,pver)   ! spec. humidity [kg/kg]
+      real(r8), intent(in) :: t(pcols,pver)   ! temperature [K]
+      real(r8), intent(in) :: p(pcols,pver)   ! pressure [hPa]
+      real(r8), intent(in) :: pf(pcols,pverp) ! pressure at interfaces [hPa]
+      real(r8), intent(in) :: zm(pcols,pver)  ! midpoint height [m]
+      real(r8), intent(in) :: zi(pcols,pverp) ! interface height [m]
+      real(r8), intent(in) :: pblh(pcols)     ! PBL height [m]
+      real(r8), intent(in) :: phis(pcols)     ! Surface geopotential [m]
+
+   !  OUTPUT arguments
+      real(r8), intent(out) :: cape(pcols)    ! convective aval. pot. energy [J/kg]
+      real(r8), intent(out) :: cin(pcols)     ! covective inhibition [J/kg]
+
+   !
+   !--------------------------Local Variables------------------------------
+   !
+      ! number of "potential CAPEs" to consider in the CAPE calculation
+      ! For this diagnostic set to the standard of 1
+      integer, parameter :: num_cin = 1
+
+      real(r8) capeten(pcols,num_cin) ! provisional value of cape
+      real(r8) tv(pcols,pver)       ! virtual temperature [K]
+      real(r8) tpv(pcols,pver)      ! virtual temperature of parcel [K]
+      real(r8) buoy(pcols,pver)     ! Buoyancy for CAPE calculations [K]
+      real(r8) neg_buoy(pcols,pver) ! "Negative" buoyancy for CIN calculations [K}
+
+      ! heights needed for diagnostic [m]
+      real(r8) :: zs(pcols), zf(pcols,pverp), z(pcols,pver)
+
+      real(r8) :: tp(pcols,pver)   ! parcel temperature [K]
+      real(r8) :: qstp(pcols,pver) ! saturation mixing ratio of parcel [kg/kg]
+      real(r8) :: tl(pcols)        ! parcel temperature at lcl [K]
+
+      integer lcl(pcols)        ! level of lifting condensation level
+      integer lel(pcols)        ! level of equilibrium level
+      integer lon(pcols)        ! level of onset of deep convection
+      integer mx(pcols)         ! level of max moist static energy
+      integer pblt(pcols)       ! integer of PBL height
+
+      ! Miscellanous 2d arrays needed for diagnostic
+      real(r8) :: a1(pcols), a2(pcols), estp(pcols), pl(pcols)
+      real(r8) :: plexp(pcols), hmax(pcols), hmn(pcols), y(pcols)
+      integer :: knt(pcols)
+      integer :: lelten(pcols,num_cin)
+
+      real(r8) :: e, tiedke_add, rgravit
+
+      ! Indicees
+      integer :: i, k, n
+
+   !
+   !-----------------------------------------------------------------------
+   !
+
+      ! Optional argument used in CAM/EAM to increase buoyancy
+      !  according to Tiedke.  For diagnostics purposes, set to zero.
+      tiedke_add = 0.0_r8
+
+      rgravit = 1._r8/gravit
+
+   !  Determine indicee of PBL height, first set up required height arrays
+      do i = 1, ncol
+         zs(i) = phis(i)*rgravit
+         zf(i,pver+1) = zi(i,pver+1) + zs(i)
+      end do
+
+      do k = 1,pver
+         do i = 1, ncol
+            z(i,k) = zm(i,k) + zs(i)
```

Comment:
@bogensch Is this to account for one of the heights being in terms of sea-level and the other being in terms of height above surface? In this case, is z supposed to be height wrt sea-level, zm is wrt to the surface, and pblh is wrt sea-level?

Reply:
Ah, yes, this section is unnecessarily complicated.  I will be pushing a commit here shortly to clean this up and with clearer comments.  In short, the index of the PBL height can be computed from standard zm and zi heights (wrt to surface).  A height computed wrt to sea-level still needs to be computed for later on in the code for the MSE.  I have verified that these new commits produce b4b CAPE and CIN diagnostics, as expected, for a SCM case over land.